### PR TITLE
Various improvements to RSS feeds and download rules

### PR DESCRIPTION
**Pull request by jwbuiter**
## Description
Made it possible to edit existing torrent feeds and download rules.
Made it possible to browse feeds and manually add torrents from them.
Added custom feed update intervals.
Added an example field to test regex matching for download rules.

## Related Issue
see #607, #425, #481, #519

## Motivation and Context
see #607, #425, #481, #519

## How Has This Been Tested?
I have added, removed and edited multiple feeds and rules and everything seems to be working as intended.

## Screenshots (if appropriate):
Too many screenshots for github:
[imgur link](https://imgur.com/a/Pqqs8)

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by N3oTraX** [comment]
Hi, how can I download your changes to review them ?

**Comment by jwbuiter** [comment]
in a new folder:

git clone -b feature/edit-feeds git@github.com:jwbuiter/flood.git

then follow normal install instructions

**Comment by jfurrow** [comment]
I'm not trying to let this PR linger, but I think this needs some UI work. I'd be happy to do this eventually, but I don't know when I'll get to it, honestly.

**Comment by N3oTraX** [comment]
Hi, feature/edit-feeds branch can't be found, I would really like to review :/

**Comment by geops22** [comment]
Being able to browse the rss feeds the same way you browse your added torrents would be optimal, e.g. like in rutorrent.

File: client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js [hunk 1]
```diff
@@ -112,15 +134,42 @@ class DownloadRulesTab extends React.Component {
...
+
+    const isMatched = (new RegExp(match, 'gi')).test(check);
+    const isExcluded = exclude !== '' && (new RegExp(exclude, 'gi')).test(check);
+    
+    if(isMatched && !isExcluded){
+      checkMatchTextbox.style.background = '#39ce83';
+    } else {
+      checkMatchTextbox.style.background = '#e95779';
+    }
+  }
```
Comment (jfurrow) [inline client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js]: We shouldn't be setting styles inline. Also, why do we need to change the styles at all?

File: client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js [hunk 2]
```diff
@@ -42,10 +44,26 @@ const MESSAGES = defineMessages({
...
     defaultMessage: 'Tags'
+  },
+  check: {
+    id: 'feeds.check',
+    defaultMessage: 'Check matching'
   }
 });
 
+const defaultRule = {
+  _id: 'new',
```
Comment (jfurrow) [inline client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js]: Shouldn't need the `_id`, this is assigned automatically by nedb.

File: server/services/feedService.js [hunk 3]
```diff
@@ -24,6 +24,17 @@ class FeedService {
     });
   }
 
+  modifyFeed(id, feed, callback) {
+    let modifiedFeed = this.feeds.find( (feed) => {
```
Comment (jfurrow) [inline server/services/feedService.js]: Use `const` and remove the whitespace at `.find( (feed`

File: client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js (hunk 1 above)
Comment (jwbuiter) [reply client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js]: It was neccesary to change the style there in order to indicate whether the given example matches the supplied patterns. 
The behaviour is shown in the last image of the imgur link. 
If the given example matches the patterns, the background will be #39ce83, otherwise #e95779.

What would be the best way to apply the styling in this case?

Comment (jfurrow) [reply client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js]: Ahh okay, I'd prefer to use CSS to style it. You can apply classes conditionally using `classnames`

**Comment by jwbuiter** [comment]
The travis build seems to be running into some problems, but only for node v9.
The log from travis is empty.
Does anybody know what might be causing this?

**Comment by jwbuiter** [comment]
Reopening the PR fixed the travis error.
@jfurrow, I have implemented your corrections

**Comment by jfurrow** [comment]
Thanks @jwbuiter and sorry for the terrible delay. I think this still needs a design pass before being accepted... are you willing to iterate on this? Or are there any designers willing to mock something up?

Here are some quick thoughts:
1. This should be an edit icon and there should be more space between the icons:
    screen shot 2018-09-20 at 9 56 42 pm
2. I'm not crazy about using the whole row for the plus sign, I'd prefer it be outside of the container (in the lighter background area) — probably as a standard `Button` component.
    screen shot 2018-09-20 at 9 57 26 pm
3. I think this form should be on the light-colored background as well:
    screen shot 2018-09-20 at 9 58 52 pm
4. I don't like that the feed selection/search are combined with the download option form, they should be separated:
    screen shot 2018-09-20 at 10 01 25 pm

With regard to `#4`, maybe there should be no download options at all — maybe each item in the feed should have a checkbox, and the user can select multiple torrents before clicking a "Download" button, at which point the "Add torrent" modal opens with the selected torrents pre-populated... that way the user will be given the opportunity to set the download options. What do you think @jwbuiter?

**Comment by jwbuiter** [comment]
Those changes look good to me. I will try to implement them soon.

Regarding `#1` , which icon do you mean when you say 'edit icon'?
Is there already such an icon in the project files(`client/src/javascript/components/icons`) or flood-ui-kit?

**Comment by bbashy** [comment]
@jwbuiter Thanks for your work on this. I know a lot of people are waiting for it. Keep it up :+1:

**Comment by jwbuiter** [comment]
@jfurrow, I hope I understood your suggestions correctly. This is what I have now:

Suggestion 1.
image

Suggestion 2.
The add button is now a normal Button component, Do you want it to have a plus sign or the word 'Add' on it? 
image

Suggestion 3.
This mostly restores the original behaviour where the form was outside of the list. There is also an indicator for which feed or rule is being edited.
image

Suggestion 4.
Pressing download opens the 'Add torrents' modal with the selected torrents.
image

File: client/src/javascript/components/general/form-elements/TextboxRepeater.js [hunk 4]
```diff
@@ -6,7 +6,7 @@ import RemoveMini from '../../icons/RemoveMini';
 
 export default class TextboxRepeater extends React.PureComponent {
   state = {
-    textboxes: [{id: 0, value: ''}],
+    textboxes: this.props.children || [{id: 0, value: ''}],
```
Comment (jfurrow) [inline client/src/javascript/components/general/form-elements/TextboxRepeater.js]: This seems weird, why is this necessary?

File: client/src/javascript/components/modals/add-torrents-modal/AddTorrentsByURL.js [hunk 5]
```diff
@@ -17,14 +18,15 @@ class AddTorrentsByURL extends React.Component {
     errors: {},
     isAddingTorrents: false,
     tags: '',
-    urlTextboxes: [{value: ''}],
+    urlTextboxes: UIStore.getActiveModal().torrents || [{id: 0, value: ''}],
```
Comment (jfurrow) [inline client/src/javascript/components/modals/add-torrents-modal/AddTorrentsByURL.js]: This breaks if `getActiveModal()` returns `null`

File: client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js [hunk 6]
```diff
@@ -149,79 +196,175 @@ class DownloadRulesTab extends React.Component {
...
+            />
+            <Textbox
+              style={{backgroundColor: '#222222'}}
+              id="check"
+              label={this.props.intl.formatMessage({
+                id: 'feeds.test.match',
+                defaultMessage: 'Check Matching.',
+              })}
+              placeholder={this.props.intl.formatMessage(MESSAGES.check)}
+            />
```
Comment (jfurrow) [inline client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js]: What does this do?

File: client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js [hunk 7]
```diff
@@ -149,79 +196,175 @@ class DownloadRulesTab extends React.Component {
...
+              id="exclude"
+              label={this.props.intl.formatMessage({
+                id: 'feeds.exclude.pattern',
+                defaultMessage: 'Exclude Pattern',
+              })}
+              placeholder={this.props.intl.formatMessage(MESSAGES.regEx)}
+              defaultValue={rule.exclude}
+            />
+            <Textbox
+              style={{backgroundColor: '#222222'}}
```
Comment (jfurrow) [inline client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js]: No inline styles

**Comment by GottZ** [comment]
@jwbuiter ping

File: client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js [hunk 8]
```diff
@@ -111,8 +130,36 @@ class DownloadRulesTab extends React.Component {
...
+      return;
+    }
+
+    const isMatched = new RegExp(match, 'gi').test(check);
+    const isExcluded = exclude !== '' && new RegExp(exclude, 'gi').test(check);
+
+    if (isMatched && !isExcluded) {
+      checkMatchTextbox.classList.add('check-match-success');
+    } else {
+      checkMatchTextbox.classList.add('check-match-failed');
```
[the remark this answers is not in the document]
Comment (GottZ) [reply client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js]: i know this goes against your standards but why avoid a class name state manager that's bound into vanilla js and glue css classes together in a stringified form?  
i'd say using the classList api is in general better practice. though i agree this code could be better. instead of `remove` and `add` one could use `toggle` and provide the conditional expression as parameter to it.

**Comment by jwbuiter** [comment]
Sorry, I'm currently working on my masters thesis so I dont have time to work on this.

**Comment by enoch85** [comment]
@jwbuiter THANK YOU! This is very nice!

File: client/src/javascript/components/modals/add-torrents-modal/AddTorrentsByURL.js (hunk 5 above)
Comment (jwbuiter) [reply client/src/javascript/components/modals/add-torrents-modal/AddTorrentsByURL.js]: I have now added a check for this.

File: client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js (hunk 7 above)
Comment (jwbuiter) [reply client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js]: I have now removed the inline styling.

I think this was left over from the previous iteration and I dont think it was doing anything.

File: client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jwbuiter) [reply client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js]: When I added a className property to the checkMatchTextbox component, that class did not show up in the resulting HTML. I think this is caused by the fact that checkMatchTextbox refers to a component which comes from flood-ui-kit, rather than a standard HTML tag.

So the classList approach was the only way i could think of to manipulate the class of this component.

**Comment by enoch85** [comment]
I'm trying to test this but the `npm install` command fails, and I cannot continue with `npm run build`. I do a git clone with this branch and then following the normal instructions.

Am I missing something or is it not working?

**Comment by jwbuiter** [comment]
I just checked and the install indeed fails, but only on the master branch.
Did you do `git checkout feature/edit-feeds` ?

**Comment by enoch85** [comment]
> Did you do git checkout feature/edit-feeds ?

I did `git clone -b feature/edit-feeds https://github.com/jwbuiter/flood.git``Isn't that correct?

**Comment by jwbuiter** [comment]
Hmm, i am now also getting a failed install due to a missing geoip database from `https://geolite.maxmind.com/download/geoip/database/GeoIPCountryCSV.zip`.

Yesterday I was able to download that file, but not anymore.

Is that also the error you are getting?

**Comment by enoch85** [comment]
@jwbuiter Check [this](https://github.com/jfurrow/flood/pull/748#issuecomment-450740170) out.

**Comment by jwbuiter** [comment]
Yes, i also had that error, but it was fixed by commit 9b6e532.

Before that fix is applied, the error is: 
```
>geoip-country-only@1.1.8 updatedb /home/jwbuiter/test/flood/node_modules/geoip-country-only
>node scripts/updatedb.js

internal/util/inspect.js:31
const types = internalBinding('types');
              ^

ReferenceError: internalBinding is not defined
    at internal/util/inspect.js:31:15
    at req_ (/home/jwbuiter/test/flood/node_modules/natives/index.js:137:5)
    at require (/home/jwbuiter/test/flood/node_modules/natives/index.js:110:12)
    at util.js:25:21
    at req_ (/home/jwbuiter/test/flood/node_modules/natives/index.js:137:5)
    at require (/home/jwbuiter/test/flood/node_modules/natives/index.js:110:12)
    at fs.js:42:21
    at req_ (/home/jwbuiter/test/flood/node_modules/natives/index.js:137:5)
    at Object.req [as require] (/home/jwbuiter/test/flood/node_modules/natives/index.js:54:10)
    at Object.<anonymous> (/home/jwbuiter/test/flood/node_modules/fstream/node_modules/graceful-fs/fs.js:1:99)
```

After that fix, the error is: 
```
> geoip-country-only@1.1.8 updatedb /home/jwbuiter/test/flood/node_modules/geoip-country-only
> node scripts/updatedb.js

Fetching  https://geolite.maxmind.com/download/geoip/database/GeoIPCountryCSV.zip
Retrieving GeoIPCountryCSV.zip .../home/jwbuiter/test/flood/node_modules/geoip-country-only/scripts/updatedb.js:111
                        console.log('ERROR'.red + ': HTTP Request Failed [%d %s]', status, https.STATUS_CODES[status]);
                                                                                                             ^

TypeError: Cannot read property '404' of undefined
    at ClientRequest.onResponse (/home/jwbuiter/test/flood/node_modules/geoip-country-only/scripts/updatedb.js:111:89)
    at Object.onceWrapper (events.js:273:13)
    at ClientRequest.emit (events.js:182:13)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (_http_client.js:556:21)
    at HTTPParser.parserOnHeadersComplete (_http_common.js:109:17)
    at TLSSocket.socketOnData (_http_client.js:442:20)
    at TLSSocket.emit (events.js:182:13)
    at addChunk (_stream_readable.js:283:12)
    at readableAddChunk (_stream_readable.js:264:11)
    at TLSSocket.Readable.push (_stream_readable.js:219:10)
```

However, this error did not occur yesterday.
Maybe the file `GeoIPCountryCSV.zip` was deleted a few hours ago

**Comment by enoch85** [comment]
Yeah, probably. Just found out that:

**Works**
https://geolite.maxmind.com/download/geoip/database/GeoLite2-Country-CSV.zip

**Fails**
https://geolite.maxmind.com/download/geoip/database/GeoIPCountryCSV.zip

**Comment by enoch85** [comment]
@jwbuiter In [this](https://github.com/jfurrow/flood/issues/751#issuecomment-451021662) comment we have the reason. Maybe update to `geoip-lite`, or even better, `geoip2-lite`

File: client/src/javascript/components/general/form-elements/TextboxRepeater.js (hunk 4 above)
Comment (jfurrow) [reply client/src/javascript/components/general/form-elements/TextboxRepeater.js]: The children of a React component should be React nodes — things that will be rendered in the DOM. This array of data is essentially the default value of the form control (`TextboxRepeater`), so it should be passed in as an explicit prop like `defaultValues`.

File: client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (jfurrow) [reply client/src/javascript/components/modals/feeds-modal/DownloadRulesTab.js]: @GottZ Because this approach breaks React's DOM abstraction.

@jwbuiter If we need to provide a custom `className` to a `flood-ui-kit` component, then the component should be updated in `flood-ui-kit`

**Comment by enoch85** [comment]
@jwbuiter `geoip-lite` is now in master. Please rebase so that it's even with master. I really want to test this PR. :)

**Comment by enoch85** [comment]
Sorry, never came around to test this due to my lack of skills setting up a npm dev enviroment. :/

Anyway, this is the last piece of the puzzle for me for moving to Flood 100% instead of ruTorrent. Any updates @jwbuiter @jfurrow ?

Thanks for your great work!

**Comment by lukaszraczylo** [comment]
I'm using this branch for a month now, looks and works okay, no issues found so far.

**Comment by billimek** [comment]
Would love to see this merged!

**Comment by jfurrow** [comment]
I'm gonna merge this into a feature branch for now
